router/sleeve.go

```diff
@@ -0,0 +1,1117 @@
+				fwd.crypto.EncDF, fwd.senderDF, fwd.maxPayload)
+
+		case special := <-specialChan:
+			err = fwd.handleSpecialFrame(special)
+
+		case _, ok := <-confirmedChan:
+			if !ok {
+				// specialChan is closed to indicate
+				// the forwarder is being closed
+				break loop
+			}
+
+			err = fwd.confirmed()
+
+		case <-timerChan(fwd.heartbeatTimer):
+			err = fwd.sendHeartbeat()
+
+		case <-timerChan(fwd.heartbeatTimeout):
+			err = fmt.Errorf("timed out waiting for UDP heartbeat")
+
+		case <-tickerChan(fwd.fragTestTicker):
+			err = fwd.sendFragTest()
+
+		case <-timerChan(fwd.pmtuTestTimeout):
+			err = fwd.handlePMTUTestFailure()
+		}
+	}
+
+	if fwd.heartbeatTimer != nil {
+		fwd.heartbeatTimer.Stop()
+	}
+	if fwd.heartbeatTimeout != nil {
+		fwd.heartbeatTimeout.Stop()
+	}
+	if fwd.fragTestTicker != nil {
+		fwd.fragTestTicker.Stop()
+	}
+	if fwd.pmtuTestTimeout != nil {
+		fwd.pmtuTestTimeout.Stop()
+	}
+
+	checkWarn(fwd.senderDF.close())
+
+	fwd.lock.RLock()
+	defer fwd.lock.RUnlock()
+	if fwd.listener != nil {
+		fwd.listener.Error(err)
+	}
+}
+
+func (fwd *sleeveForwarder) aggregateAndSend(frame aggregatorFrame,
+	aggChan <-chan aggregatorFrame, enc Encryptor, sender udpSender,
+	limit int) error {
+	// Give up after processing N frames, to avoid starving the
+	// other activities of the forwarder goroutine.
+	i := 0
+
+	for {
+		// Adding the first frame to an empty buffer
+		if !fits(frame, enc, limit) {
+			log.Println(fwd.logPrefix(), "Dropping too big frame during forwarding: frame len", len(frame.frame), ", limit ", limit)
+			return nil
+		}
+
+		for {
+			enc.AppendFrame(frame.src, frame.dst, frame.frame)
+			i++
+
+			gotOne := false
+			if i < 100 {
+				select {
+				case frame = <-aggChan:
+					gotOne = true
+				default:
+				}
+			}
+
+			if !gotOne {
+				return fwd.flushEncryptor(enc, sender)
+			}
+
+			// Accumulate frames until doing so would
+			// exceed the PMTU.  Even in the non-DF case,
+			// it doesn't seem worth adding a frame where
+			// that would lead to fragmentation,
+			// potentially delaying or risking other
+			// frames.
+			if !fits(frame, enc, fwd.maxPayload) {
+				break
+			}
+		}
+
+		if err := fwd.flushEncryptor(enc, sender); err != nil {
+			return err
+		}
+	}
+}
+
+func fits(frame aggregatorFrame, enc Encryptor, limit int) bool {
+	return enc.TotalLen()+enc.FrameOverhead()+len(frame.frame) <= limit
+}
+
+func (fwd *sleeveForwarder) flushEncryptor(enc Encryptor,
+	sender udpSender) error {
+	msg, err := enc.Bytes()
+	if err != nil {
+		return err
+	}
+
+	return fwd.processSendError(sender.send(msg, fwd.remoteAddr))
+}
+
+func (fwd *sleeveForwarder) sendSpecial(enc Encryptor, sender udpSender,
+	data []byte) error {
+	enc.AppendFrame(fwd.sleeve.localPeerBin, fwd.remotePeerBin, data)
+	return fwd.flushEncryptor(enc, sender)
+}
+
+func (fwd *sleeveForwarder) handleSpecialFrame(special specialFrame) error {
+	if special.sender == nil {
+		return fwd.handleControlMsg(special.frame)
```

Comment:
It's not the fields I'd be worried about.  I haven't studied it first-hand, but my understanding of golang's select implementation is that on each pass through the select, it re-evaluates all the channel expressions into an array and then calls some run-time function to check if any are ready, and sleep if not.  So adding an extra case, even one that rarely fires, has a performance impact.  For that reason I was concerned about the performance implications of that big select in a performance-critical area.  Fortunately, when I measured the end-result, it seems to match mainline weave.  But the concern hasn't completely gone away.

If one wanted to address that performance concern properly, then it would be done by merging the various channels, with an auxiliary goroutine to fan-in the various timer events onto the merged channel.  One way to look at use of the `specialChan` for both special UDP packets and control messages is as that kind of merging, in a case where it happens to be straightforward to do so.